src/main/java/com/avispl/symphony/dal/avdevices/power/middleatlantic/nexsys/MiddleAtlanticUPSCommunicator.java

```diff
@@ -0,0 +1,962 @@
+			String response = sendCommand(command);
+			if (UPSConstant.FAIL_RESPONSE.equals(response)) {
+				throw new IllegalArgumentException("Error when send outlet status command. The request is rejected");
+			}
+		} catch (Exception e) {
+			throw new IllegalArgumentException(String.format("Can't control %s with %s value. ", propertyName, UPSConstant.NUMBER_ONE.equals(value) ? UPSConstant.ON : UPSConstant.OFF) + e.getMessage(), e);
+		}
+	}
+
+	/**
+	 * Sends a command to cycle (toggle) the status of an outlet and handles the response.
+	 *
+	 * @param propertyName The name of the outlet property to cycle.
+	 * @throws IllegalArgumentException If an error occurs while sending the outlet cycle command or if the request is rejected.
+	 * @throws InterruptedException If the thread sleep is interrupted.
+	 */
+	private void sendCycleCommand(String propertyName) {
+		try {
+			String command = UPSControlCommand.OUTLET_CYCLE_COMMAND;
+			command = command.replace("$", propertyName.replace(UPSConstant.CYCLE_OUTLET, UPSConstant.EMPTY));
+			String response = sendCommand(command);
+			if (UPSConstant.FAIL_RESPONSE.equals(response)) {
+				throw new IllegalArgumentException("Error when send outlet cycle command. The request is rejected");
+			}
+		} catch (Exception e) {
+			throw new IllegalArgumentException(String.format("Can't control %s. ", propertyName) + e.getMessage(), e);
+		}
+	}
+
+	/**
+	 * Sends a command to set the replacement date of a specific property and handles the response.
+	 *
+	 * @param propertyName The name of the property for which the replacement date is being set.
+	 * @param lastDate The last replacement date value to be set.
+	 * @param nextDate The next replacement date value to be set.
+	 * @throws IllegalArgumentException If an error occurs while sending the replacement date command or if the request is rejected.
+	 */
+	private void sendReplacementDateCommand(String propertyName, String lastDate, String nextDate) {
+		try {
+			String command = UPSControlCommand.REPLACEMENT_DATE_COMMAND;
+			command = command.replace("$1", lastDate).replace("$2", nextDate);
+			String response = sendCommand(command);
+			if (UPSConstant.FAIL_RESPONSE.equals(response)) {
+				throw new IllegalArgumentException("Error when send replacement date command. The request is rejected");
+			}
+		} catch (Exception e) {
+			throw new IllegalArgumentException(String.format("Can't control %s. ", propertyName) + e.getMessage(), e);
+		}
+	}
+
+	/**
+	 * Update cache device data
+	 *
+	 * @param cacheMapOfPropertyNameAndValue the cacheMapOfPropertyNameAndValue are map key and value of it
+	 * @param property the key is property name
+	 * @param value the value is String value
+	 */
+	private void updateCachedDeviceData(Map<String, String> cacheMapOfPropertyNameAndValue, String property, String value) {
+		cacheMapOfPropertyNameAndValue.remove(property);
+		cacheMapOfPropertyNameAndValue.put(property, value);
```

Comment:
not sure if you need to remove the value, since put() will replace it if it already exists